### PR TITLE
S6

### PR DESCRIPTION
**Pull request by Elyaszc**
comment section added, functionality added, saves to DB fine, shows user name of comment poster etc.

### CODE REVIEW

File: singlePost.php [hunk 1]
```diff
@@ -39,5 +42,55 @@
...
+    </section>
+    <?php
+}
+?>
+
+
+<section class="container md:w-1/2 mx-auto mt-5 mb-10">
+    <?php
+    foreach ($comments as $comment) : ?>
+    <div class="p-8 border border-solid rounded-md bg-slate-200">
```
Comment (ashleycoles) [inline singlePost.php]: Please can you add a `mb-2` class to this div

File: src/Entities/Comment.php [hunk 2]
```diff
@@ -8,13 +8,16 @@
...
-    public function __construct(int $id, int $authorid, int $blogid, string $content, string $timeStamp)
+    public string $username;
+
+    public function __construct(int $id, int $authorid, int $blogid, string $content, string $timeStamp, string $username)
     {
         $this->id = $id;
         $this->authorid = $authorid;
         $this->blogid = $blogid;
         $this->content = $content;
         $this->timeStamp = gmdate("d/m/y", strtotime($timeStamp));
```
Comment (ashleycoles) [inline src/Entities/Comment.php:19]: Can we get a unit test on this please, just to make sure that the timeStamp format works correctly. Just like you did with the blog test

File: singlePost.php [hunk 3]
```diff
@@ -39,5 +42,55 @@
...
+                } else
+                {
+                    $authorid = $_SESSION['userid'];
+                    $blogid = $_GET['id'];
+                    $timestamp = date('Y-m-d H:i:s');
+                    $commentModel->addComment($authorid, $blogid , $content, $timestamp);
+                    echo "<div style='display: flex; align-items: center; margin-top: -5px; color: green'><p>Comment added successfully!</p></div>";
+                }
+            }
+            ?>
```
Comment (ashleycoles) [inline singlePost.php]: Can this code move please, it's better to keep heavy logic like this at the top of the file. 

You will of course need to make sure the error messages are displaying in the correct place though.

By putting this code before `$comments = $commentModel->getAllComments($_GET['id']);` you'll find that the comment loads straight away as well, which is nice!

File: src/Models/CommentModel.php [hunk 4]
```diff
@@ -8,4 +8,43 @@ public function __construct(PDO $db)
...
+    /**
+     * @return Comment[];
+     */
+    public function hydrateComments(array $data): array | false
+    {
+        $comments = [];
+        foreach ($data as $comment)
+        {
+        $comments[] =  new Comment($comment['id'], $comment['authorid'],
+        $comment['blogid'], $comment['content'], $comment['timestamp'], $comment['username']);
```
Comment (ashleycoles) [inline src/Models/CommentModel.php]: WRong indetation. 

When breaking arguments onto multiple lines, the correct style is like this:

```
$comments[] = new Comment(
    $comment['id],
    $comment['authorId'],
    $comment['blogid']
);

```

File: src/Models/CommentModel.php [hunk 5]
```diff
@@ -8,4 +8,43 @@ public function __construct(PDO $db)
...
+       `comments`.`content`, `comments`.`timestamp`, `users` . `username` FROM `comments` LEFT JOIN `users` 
+        ON `comments` . `authorid` = `users` . `id` WHERE `comments` . `blogid` = :id ORDER BY `timestamp` DESC;');
+        $query ->execute([':id'=>$id]);
+        $data = $query->fetchAll();
+        return $this->hydrateComments($data);
+    }
+    /**
+     * @return Comment[];
+     */
+    public function hydrateComments(array $data): array | false
```
Comment (ashleycoles) [inline src/Models/CommentModel.php]: This method can ever return false

File: src/Models/CommentModel.php [hunk 6]
```diff
@@ -8,4 +8,43 @@ public function __construct(PDO $db)
...
+            'authorid'=> $authorid,
+            'blogid'=> $blogid,
+            'content'=> $content,
+            'timestamp'=> $timeStamp
+        ]);
+    }
+    /**
+     * @return Comment[];
+     */
+    public function getAllComments(int $id): array | false
```
Comment (ashleycoles) [inline src/Models/CommentModel.php]: This can't return false

File: singlePost.php [hunk 7]
```diff
@@ -39,5 +42,55 @@
...
+        <form class="p-8 border border-solid rounded-md bg-slate-200" method="post">
+            <div class="mb-5">
+                <label class="mb-3 block" for="content">Comment:</label>
+                <textarea class="w-full" id="content" name="content" rows="5"></textarea>
+            </div>
+            <?php
+            if(isset($_POST['content']))
+            {
+                // inputed special characters won't be interpreted as HTML tags/entities.
+                $content = htmlspecialchars($_POST['content'], ENT_QUOTES, 'UTF-8');
```
Comment (ashleycoles) [inline singlePost.php]: As an extra safety, can you run content through strip_tags() as well - it'll remove any HTML tags in the content

**Review by ashleycoles — CHANGES_REQUESTED** [review]

File: singlePost.php (hunk 1 above)
Comment (Elyaszc) [reply singlePost.php]: done

File: src/Entities/Comment.php (hunk 2 above)
Comment (Elyaszc) [reply src/Entities/Comment.php:19]: done

File: singlePost.php (hunk 3 above)
Comment (Elyaszc) [reply singlePost.php]: done

File: src/Models/CommentModel.php (hunk 4 above)
Comment (Elyaszc) [reply src/Models/CommentModel.php]: done

File: src/Models/CommentModel.php (hunk 5 above)
Comment (Elyaszc) [reply src/Models/CommentModel.php]: done

File: src/Models/CommentModel.php (hunk 6 above)
Comment (Elyaszc) [reply src/Models/CommentModel.php]: done

File: singlePost.php (hunk 7 above)
Comment (Elyaszc) [reply singlePost.php]: done

**Review by ashleycoles — APPROVED** [review]

**Pull request merged**